[iOS] Basic Vulkan/Metal Support

naithar opened the pull request:
This PR implements basic Vulkan support on iOS.

Minimal iOS version switched to `11.0` for iOS device and to `13.0` for iOS Simulator. Could probably be lowered back on GLES2 implementation. Also MoltenVK seems to work with iOS 9.
Uses static `.a` MoltenVK library. This allows testing on simulator (with library supporting `x86_64` architecture) and also results in smaller app size (30MB vs 33MB with `.framework`).

Patches:
- `Vulkan Memory Allocator` was patched to support running on iOS. Related Issue: https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator/issues/102
- Scrolling wasn't working for touch devices
- Release build was crashing on touch

Issues that I've found:
- Some shaders failed to build on some devices. Failed on iPhone 6 and iPad Pro 2019. On iPhone 6 it results in stuck application, while on iPad it only displays blacks screen but game loop (input/audio/keyboard) is working. On iPhone XS and iPhone 8 compilation finished successfully. 
  [Logs taken from devices](https://gist.github.com/naithar/6a349dc4001c7f75a1a05ef3dd8ebf74). 
  Possibly related issue in **MoltenVK** repo: https://github.com/KhronosGroup/MoltenVK/issues/448. 
  Switching to `master` branch of **MoltenVK** didn't help.
  Seems to be more connected with how **SPIRV-Cross** translates shaders then **MoltenVK**
  Samples of **MoltenVK** SDK work without any issue.

- **MoltenVK** source that will be used for `.a` library should contain a fix linked to this issue: https://github.com/KhronosGroup/MoltenVK/issues/950. Otherwise application will not pass AppStore checks on Private API.

- ~~`ClassDB::bind_method(D_METHOD(...))` causes crash on `Main::cleanup()` at `MethodBind1R<Variant, String>::~MethodBind1R()`. 
  Crashes that was happening from bindings of `game_center` and `app_store` modules was worked around by switching from `Variant` to simpler types like `Array`, `Dictionary`. 
  For `icloud` module's methods `Array set_key_values(Dictionary p_params)` and `Variant get_key_value(Variant p_param)` nothing seems to be working.~~

naithar committed:
GUI ScrollBar: possible fix for scrolling

Use of unmodified value returned by 'screen_is_touchscreen' to be used in determening if scroll bar should be scrolled

naithar committed:
Thirdparty Vulkan: patch VMA to fix assets

Applies VMA master branch patch that removes incorrect asserts:
issue: https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator/issues/102
patch: https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator/commit/39aeff7a434801c5f8a2432b9544a2165e63e697

naithar committed:
iOS SCons: update iOS minimal version

iOS 11 for iOS device
iOS 13 for iOS Simulator

naithar committed:
Modules: update modules to be built for iOS

Using 'available' checks to fix deprecation compilation errors
Additional checks for simulator

naithar committed:
Core Variant: switch from 'real' to 'float' to allow building with NEED_LONG_INT

naithar committed:
iOS: Vulkan support

Implemented Vulkan Support.
Use DisplayServer for rendering and input handling
Use single view for rendering in both GLES2 (not supported yet) and Vulkan
Use @available checks where it's required (otherwise compiler would fail compilation)
Simulator checks

naithar committed:
iOS Vulkan: copy icd to app

naithar commented:
> * `ClassDB::bind_method(D_METHOD(...))` causes crash on `Main::cleanup()` at `MethodBind1R<Variant, String>::~MethodBind1R()`. Binding crash from `game_center` and `app_store` was worked around by switching from `Variant` to simpler type like `Array`, `Dictionary`. For `Array set_key_values(Dictionary p_params)` and `Variant get_key_value(Variant p_param)` nothing seems to work.

I guess this would require a separate issue once I can debug it correctly.

naithar commented:
> Uses `.framework` built from `master` of **MoltenVK** repo for now.

Also Vulkan seems be be searching for dynamic library, so it's either `.framework` or `.dylib` (which is not allowed by AppStore). 
Not sure if it's on Godot side only, but MoltenVK Sample project runs with static `.a` library just fine.

bruvzg commented:
>Also Vulkan seems be be searching for dynamic library, so it's either .framework or .dylib (which is not allowed by AppStore).
Not sure if it's on Godot side only, but MoltenVK Sample project runs with static .a library just fine.

Godot uses dynamic Vulkan loader, MVK samples link to MVK driver directly.

On macOS static MoltenVK linking without loader is implemented with `use_static_mvk` build option, probably the same should be used on iOS (and enabled by default, since there no benefits in the loader on iOS anyway).

naithar commented:
> On macOS static MoltenVK linking without loader is implemented with `use_static_mvk` build option, probably the same should be used on iOS (and enabled by default, since there no benefits in the loader on iOS anyway).

Using `use_static_mvk` for iOS results in errors:
```
./drivers/vulkan/vulkan_context.h:40:10:{40:10-40:27}: fatal error: 'vulkan/vulkan.h' file not found [1]
 #include <vulkan/vulkan.h>
          ^~~~~~~~~~~~~~~~~
```
This gets fixed by adding `env.Prepend(CPPPATH=["#thirparty/vulkan/include"])` not sure why it doesn't happen on any other flag combination. But in the end iOS app compilation fails with missing symbols.
I guess I'm missing something =/

Using `builtin_vulkan=yes` instead seems to work for using `.a` file and passes AppStore checks, but resulting size doesn't change this much.

bruvzg commented:
> Uses .framework built from master of MoltenVK repo for now.

Pre-built MVK binary probably should not be included into the repo, probably MoltenVK build instructions should be added to `godot-docs -> development/compiling/compiling_for_ios` instead (and it should be for universal x86_64 + arm64 to support Simulator).

---

There's unused and not existing `ios1/vulkan` library referenced in the Xcode project.

---

> Minimal iOS version switched to 11.0 since MoltenVK requires this version. Could probably be lowered back on GLES2 implementation.

`CAMetalLayer` is used and it's not available in 11.0, min version should be bumped to 13.0.
(MoltenVK should support some other backing layer for older versions, or it's dropped and documentation was not updated).

---

There are few mismatched signed/unsigned loop counters preventing it from building with `warnings=extra`:

<details>
  <summary>Click to expand</summary>

```diff
diff --git a/platform/iphone/game_center.mm b/platform/iphone/game_center.mm
index 0f73f536bd..f9ec271cfa 100644
--- a/platform/iphone/game_center.mm
+++ b/platform/iphone/game_center.mm
@@ -202,7 +202,7 @@ void GameCenter::request_achievement_descriptions() {
 			Array hidden;
 			Array replayable;
 
-			for (int i = 0; i < [descriptions count]; i++) {
+			for (NSUInteger i = 0; i < [descriptions count]; i++) {
 				GKAchievementDescription *description = [descriptions objectAtIndex:i];
 
 				const char *str = [description.identifier UTF8String];
@@ -250,7 +250,7 @@ void GameCenter::request_achievements() {
 			PackedStringArray names;
 			PackedFloat32Array percentages;
 
-			for (int i = 0; i < [achievements count]; i++) {
+			for (NSUInteger i = 0; i < [achievements count]; i++) {
 				GKAchievement *achievement = [achievements objectAtIndex:i];
 				const char *str = [achievement.identifier UTF8String];
 				names.push_back(String::utf8(str != NULL ? str : ""));
diff --git a/platform/iphone/godot_iphone.mm b/platform/iphone/godot_iphone.mm
index dd97ae6aa5..e15d18539d 100644
--- a/platform/iphone/godot_iphone.mm
+++ b/platform/iphone/godot_iphone.mm
@@ -63,7 +63,7 @@ int add_cmdline(int p_argc, char **p_args) {
 		return p_argc;
 	}
 
-	for (int i = 0; i < [arr count]; i++) {
+	for (NSUInteger i = 0; i < [arr count]; i++) {
 		NSString *str = [arr objectAtIndex:i];
 		if (!str) {
 			continue;
diff --git a/platform/iphone/icloud.mm b/platform/iphone/icloud.mm
index aae8278524..6d1f8daa44 100644
--- a/platform/iphone/icloud.mm
+++ b/platform/iphone/icloud.mm
@@ -163,7 +163,7 @@ NSObject *variant_to_nsobject(Variant v) {
 		NSMutableDictionary *result = [[[NSMutableDictionary alloc] init] autorelease];
 		Dictionary dic = v;
 		Array keys = dic.keys();
-		for (unsigned int i = 0; i < keys.size(); ++i) {
+		for (int i = 0; i < keys.size(); ++i) {
 			NSString *key = [[[NSString alloc] initWithUTF8String:((String)(keys[i])).utf8().get_data()] autorelease];
 			NSObject *value = variant_to_nsobject(dic[keys[i]]);
 
@@ -177,7 +177,7 @@ NSObject *variant_to_nsobject(Variant v) {
 	} else if (v.get_type() == Variant::ARRAY) {
 		NSMutableArray *result = [[[NSMutableArray alloc] init] autorelease];
 		Array arr = v;
-		for (unsigned int i = 0; i < arr.size(); ++i) {
+		for (int i = 0; i < arr.size(); ++i) {
 			NSObject *value = variant_to_nsobject(arr[i]);
 			if (value == NULL) {
 				//trying to add something unsupported to the array. cancel the whole array
@@ -215,7 +215,7 @@ Array ICloud::set_key_values(Dictionary p_params) {
 
 	Array error_keys;
 
-	for (unsigned int i = 0; i < keys.size(); ++i) {
+	for (int i = 0; i < keys.size(); ++i) {
 		String variant_key = keys[i];
 		Variant variant_value = p_params[variant_key];
 diff --git a/modules/camera/camera_ios.mm b/modules/camera/camera_ios.mm
index b917eb537a..609d55e79c 100644
--- a/modules/camera/camera_ios.mm
+++ b/modules/camera/camera_ios.mm
@@ -163,8 +163,8 @@
 
 		{
 			// do Y
-			int new_width = CVPixelBufferGetWidthOfPlane(pixelBuffer, 0);
-			int new_height = CVPixelBufferGetHeightOfPlane(pixelBuffer, 0);
+			size_t new_width = CVPixelBufferGetWidthOfPlane(pixelBuffer, 0);
+			size_t new_height = CVPixelBufferGetHeightOfPlane(pixelBuffer, 0);
 			//			int _bytes_per_row = CVPixelBufferGetBytesPerRowOfPlane(pixelBuffer, 0);
 
 			if ((width[0] != new_width) || (height[0] != new_height)) {
@@ -184,8 +184,8 @@
 
 		{
 			// do CbCr
-			int new_width = CVPixelBufferGetWidthOfPlane(pixelBuffer, 1);
-			int new_height = CVPixelBufferGetHeightOfPlane(pixelBuffer, 1);
+			size_t new_width = CVPixelBufferGetWidthOfPlane(pixelBuffer, 1);
+			size_t new_height = CVPixelBufferGetHeightOfPlane(pixelBuffer, 1);
 			//			int bytes_per_row = CVPixelBufferGetBytesPerRowOfPlane(pixelBuffer, 1);
 
 			if ((width[1] != new_width) || (height[1] != new_height)) {
```

</details>

And multiple deprecated APIs used (unrelated to this PR), but builds with `warnings=extra` and `CXXFLAGS=-Wno-deprecated-declarations`.

naithar commented:
> Pre-built MVK binary probably should not be included into the repo, probably MoltenVK build instructions should be added to `godot-docs -> development/compiling/compiling_for_ios` instead (and it should be for universal x86_64 + arm64 to support Simulator).

Thought of that too, should be an easy fix :)

> `CAMetalLayer` is used and it's not available in 11.0, min version should be bumped to 13.0.
> 
> (MoltenVK should support some other backing layer for older versions, or it's dropped and documentation was not updated).

`CAMetalLayer` seems to be requiring iOS 13 only for Simulator, but on device it's present since iOS 8. I'll look into it.


Also iOS release build crashes when processing touches at `Viewport::_make_input_local`, not sure if it's related to this PR.

naithar committed:
Core Callable: fixed variant call caster

Use same call to VariantCaster in release build as used in VariantCasterAndValidate::call method

naithar committed:
iOS Export: MoltenVK framework for Vulkan support

naithar commented:
I've switch usage of `.framework` to `.a`. 
Fixed deprecation warnings for modules.
Added a workaround for Simulator.
Also fixed some things in `core`.

naithar commented on core/callable_method_pointer.h:
```diff
@@ -131,7 +131,7 @@ void call_with_variant_args_helper(T *p_instance, void (T::*p_method)(P...), con
 	(p_instance->*p_method)(VariantCasterAndValidate<P>::cast(p_args, Is, r_error)...);
 #else
-	(p_instance->*p_method)(VariantCaster<P>::cast(p_args[Is])...);
+	(p_instance->*p_method)(VariantCaster<P>::cast(*p_args[Is])...);
```
Now it does the same thing that happens in `VariantCasterAndValidate<P>::cast` after checks:
```
VariantCaster<T>::cast(*p_args[p_arg_idx]);
```

naithar commented on drivers/vulkan/SCsub:
```diff
@@ -22,6 +22,17 @@ if env["platform"] == "android":
     thirdparty_dir = "#thirdparty/vulkan"
     vma_sources = [thirdparty_dir + "/android/vk_mem_alloc.cpp"]
     env_thirdparty.add_source_files(env.drivers_sources, vma_sources)
+if env["platform"] == "iphone":
```
VMA is required for iPhone for Vulkan to work

naithar committed:
iOS SCons: static Vulkan binary usage

Add VMA to iphone platform Use linkflag for iphone building to enforce static linking. Works fine with dynamic '.framework' library
Updated xcode project to use '.a' static library

naithar commented:
> * Some shaders failed to build on some devices. Failed on iPhone 6 and iPad Pro 2019. On iPhone 6 it results in stuck application, while on iPad it only displays blacks screen but game loop (input/audio/keyboard) is working. On iPhone XS and iPhone 8 compilation finished successfully. [Logs of failed shaders](https://gist.github.com/naithar/5d82866df7d8cdf0ef8916e6b480343b).
>   Possibly related issue in **MoltenVK** repo: [KhronosGroup/MoltenVK#448](https://github.com/KhronosGroup/MoltenVK/issues/448).
>   Switching to `master` branch of **MoltenVK** didn't help.
>   Seems to be more connected with how **SPIRV-Cross** translates shaders then **MoltenVK**
>   Samples of **MoltenVK** SDK work without any issue.

I tried to find shader causing the problem. Upon removing `copy.glsl` and code that uses it, iPhone 6 and iPad got an error:
```
[mvk-info] Created 3 swapchain images with initial size (750, 1334).
2020-07-17 14:08:22.565679+0300 godot_app[527:101855] ERROR: Unable to create DisplayServer, all display drivers failed.
2020-07-17 14:08:22.566851+0300 godot_app[527:101855]    at: setup2 (main/main.cpp:1304) - Unable to create DisplayServer, all display drivers failed.
ERROR: Unable to create DisplayServer, all display drivers failed.
   at: setup2 (main/main.cpp:1304) - Unable to create DisplayServer, all display drivers failed.
[mvk-info] Created 3 swapchain images with initial size (750, 1334).
2020-07-17 14:08:22.573511+0300 godot_app[527:101855] ERROR: Condition "!_start_success" is true. Returning: false
2020-07-17 14:08:22.574230+0300 godot_app[527:101855]    at: start (main/main.cpp:1539) - Condition "!_start_success" is true. Returning: false
ERROR: Condition "!_start_success" is true. Returning: false
   at: start (main/main.cpp:1539) - Condition "!_start_success" is true. Returning: false
```

But iPhone 8 and iPhone XS worked fine.

naithar commented:
> * `ClassDB::bind_method(D_METHOD(...))` causes crash on `Main::cleanup()` at `MethodBind1R<Variant, String>::~MethodBind1R()`.
>   Crashes that was happening from bindings of `game_center` and `app_store` modules was worked around by switching from `Variant` to simpler types like `Array`, `Dictionary`.
>   For `icloud` module's methods `Array set_key_values(Dictionary p_params)` and `Variant get_key_value(Variant p_param)` nothing seems to be working.

Got fixed after rebase.

> I tried to find shader causing the problem. Upon removing `copy.glsl` and code that uses it, iPhone 6 and iPad got an error:

Fixed a `Unable to create DisplayServer` error, but shader compilation still fails. Updated logs in opening post.

naithar commented:
Updated [Logs](https://gist.github.com/naithar/6a349dc4001c7f75a1a05ef3dd8ebf74) with recent findings. 

Disabling all shaders allowed iPhone6 to get into main loop and perform gesture interactions, play music, display alerts.
But both iPhone6 and iPad got black screen, showing nothing.
iPhoneXS was running fine, but got some error messages.

Building MoltenVK with enabled `MVK_CONFIG_FULL_IMAGE_VIEW_SWIZZLE` gave the same results for all three devices - black screen, but working interactions. (`Removed Shaders v2.md` in logs)

akien-mga commented:
@naithar Since we migrated from Travis to GitHub Actions and I removed the `.travis.yml` file, there's no CI build for iOS anymore.

I added a workflow for iOS builds on GitHub Actions on top of this PR: https://github.com/akien-mga/godot/tree/naithar-feature/ios-moltenVK

May I force push my changes to this PR to update it?

I also amended your commit logs to make sure that there's an empty line between commit titles and commit description, as that's the intended format for multiline commit messages.

bruvzg commented:
> Building MoltenVK with enabled MVK_CONFIG_FULL_IMAGE_VIEW_SWIZZLE

On macOS, it's enabled by setting env variable on app start, same should work on iOS too.

https://github.com/godotengine/godot/blob/93b50a62e333cd05702fb749ebcdc18127abe01b/platform/osx/godot_main_osx.mm#L39-L42

akien-mga commented on modules/arkit/arkit_interface.mm:
```diff
@@ -162,37 +172,42 @@
-		/* important, NOT scaled to world_scale !! */
-		arr.push_back(transform);
-	}
+	if (@available(iOS 11, *)) {
```
```suggestion
	if (!@available(iOS 11, *)) {
		return Array();
	}

	Array arr;
	...
```
That saves an extra indentation and a lot of diff. Same for other methods which follow the same pattern (early return instead of big `if`).

naithar replied:
Using negative check for `available` might not work (I think I have tried that), but I'll look into it.

bruvzg replied:
Checks like `if (@available(..) || something)` do not compile, it's probably same with the `not`.

akien-mga commented on modules/arkit/arkit_interface.mm:
```diff
@@ -286,25 +305,28 @@
 
 Size2 ARKitInterface::get_render_targetsize() {
-	_THREAD_SAFE_METHOD_
+	//	_THREAD_SAFE_METHOD_
```
Why is this needed?

Style wise, it should be:
```suggestion
	//_THREAD_SAFE_METHOD_
```

naithar replied:
`_THREAD_SAFE_METHOD` is used in old ARKit implementation, but in `master` gives a `Mutex` error on iOS, so I've commented it. 
Will fix styling :)

akien-mga commented on modules/arkit/arkit_interface.mm:
```diff
@@ -443,8 +465,18 @@
+
+				if (@available(iOS 13, *)) {
+					orientation = [UIApplication sharedApplication].delegate.window.windowScene.interfaceOrientation;
+#if TARGET_OS_SIMULATOR
+				}
+#else
+				} else {
+					orientation = [[UIApplication sharedApplication] statusBarOrientation];
+				}
+#endif
```
Another option:
```suggestion
#ifndef TARGET_OS_SIMULATOR
				} else {
					orientation = [[UIApplication sharedApplication] statusBarOrientation];
#endif
				}
```
But it might be a bit less readable :)

akien-mga commented on modules/camera/camera_ios.mm:
```diff
@@ -158,14 +158,25 @@ - (void)captureOutput:(AVCaptureOutput *)captureOutput didOutputSampleBuffer:(CM
-			int _bytes_per_row = CVPixelBufferGetBytesPerRowOfPlane(pixelBuffer, 0);
+			size_t new_width = CVPixelBufferGetWidthOfPlane(pixelBuffer, 0);
+			size_t new_height = CVPixelBufferGetHeightOfPlane(pixelBuffer, 0);
+			//			int _bytes_per_row = CVPixelBufferGetBytesPerRowOfPlane(pixelBuffer, 0);
```
If this is unused code, it's better to remove it.

akien-mga commented on platform/iphone/display_layer.mm:
```diff
@@ -0,0 +1,192 @@
+
+#import "display_layer.h"
+
+#import <OpenGLES/EAGL.h>
+#import <OpenGLES/ES1/gl.h>
+#import <OpenGLES/ES1/glext.h>
```
We usually include system headers last. See https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes

akien-mga commented on platform/iphone/display_layer.mm:
```diff
@@ -0,0 +1,192 @@
+#import <UIKit/UIKit.h>
+
+#include "main/main.h"
+#include "os_iphone.h"
```
Included twice. Grouping them all should let clang-format find and fix this.

akien-mga commented on platform/iphone/display_layer.mm:
```diff
@@ -0,0 +1,192 @@
+#include "main/main.h"
+#include "os_iphone.h"
+
+#import "GameController/GameController.h"
```
Should use `<GameController/GameController.h>` format as it's a system header.

akien-mga commented on platform/iphone/display_server_iphone.h:
```diff
@@ -0,0 +1,201 @@
+#include "drivers/vulkan/rendering_device_vulkan.h"
+#include "servers/rendering/rasterizer_rd/rasterizer_rd.h"
+
+#include "vulkan_context_iphone.h"
+#import <QuartzCore/CAMetalLayer.h>
+#include <vulkan/vulkan_metal.h>
```
```suggestion
#include "vulkan_context_iphone.h"

#import <QuartzCore/CAMetalLayer.h>
#include <vulkan/vulkan_metal.h>
```

akien-mga commented on platform/iphone/display_server_iphone.mm:
```diff
@@ -0,0 +1,753 @@
+
+#include "display_server_iphone.h"
+
+#include "core/io/file_access_pack.h"
+#include "core/project_settings.h"
+
+#include "ios.h"
+#include "os_iphone.h"
+
+#import "app_delegate.h"
+#import "godot_view.h"
+#import "view_controller.h"
+#import <Foundation/Foundation.h>
+#import <sys/utsname.h>
```
```suggestion
#import "app_delegate.h"
#include "core/io/file_access_pack.h"
#include "core/project_settings.h"
#import "godot_view.h"
#include "ios.h"
#include "os_iphone.h"
#import "view_controller.h"

#import <Foundation/Foundation.h>
#import <sys/utsname.h>
```
(or whatever order `clang-format` will impose, if it treats `#import` and `#include` differently)

akien-mga commented:
I didn't review the iOS code in 4f99d5b4a761166e83c09f16dfca300a584240e9 in depth, but the rest looks good. That's a tremendous amount of work :O

I'm happy to get this merged soon so that we can have CI going for iOS, and we can interate on actually making things *work* in follow-up PRs.

My GitHub Actions workflow seems to be working for iOS: https://github.com/akien-mga/godot/runs/906339047

naithar commented:
> May I force push my changes to this PR to update it?

Sure, I have no problem with that :)

> On macOS, it's enabled by setting env variable on app start, same should work on iOS too.

Well, I see no problem with setting this variable this way other than that using it breaks rendering on every device.


I will look at review comments a bit later and make required changes :)

naithar commented:
Fixed headers ordering with `clang-format` and styling issues.
Changed `TARGET_OS_SIMULATOR` checks a little. 
`@available` checks couldn't be changed - would give `@available does not guard availability here; use if (@available) instead` error.

Enabling `MVK_CONFIG_FULL_IMAGE_VIEW_SWIZZLE` via env variable allowed Godot's project to correctly run on iPad, while still giving a shader compilation error: `XPC_ERROR_CONNECTION_INTERRUPTED`.
iPhone 6 gives a `VK_ERROR_FEATURE_NOT_PRESENT: Device Apple A8 GPU does not support arrays of textures.` error.

akien-mga committed:
CI: Add iOS build on GitHub Actions

akien-mga commented:
I updated the PR branch to add the GitHub Action CI for iOS.

If @bruvzg is fine with the current state, this should be good to merge.

akien-mga commented:
Thanks!
